packages/shared/lib/services/endUser.service.ts

```diff
@@ -191,74 +200,107 @@ export async function upsertEndUser(
     {
         account,
         environment,
-        endUserPayload,
-        organization
+        connection,
+        endUser
     }: {
         account: DBTeam;
         environment: DBEnvironment;
-        endUserPayload: ConnectSessionInput['end_user'];
-        organization?: ConnectSessionInput['organization'];
+        connection: Pick<DBConnection, 'end_user_id'>;
+        endUser: InternalEndUser;
     }
 ): Promise<Result<EndUser, EndUserError>> {
-    // Check if the endUser exists in the database
-    const endUserRes = await getEndUser(db, { endUserId: endUserPayload.id, accountId: account.id, environmentId: environment.id }, { forUpdate: true });
-
-    // --- Create end user if it doesn't exist yet
-    if (endUserRes.isErr()) {
-        if (endUserRes.error.code !== 'not_found') {
-            return endUserRes;
-        }
-
-        const createdEndUser = await createEndUser(db, {
-            endUserId: endUserPayload.id,
-            email: endUserPayload.email || null,
-            displayName: endUserPayload.display_name || null,
-            organization: organization?.id
-                ? {
-                      organizationId: organization.id,
-                      displayName: organization.display_name || null
-                  }
-                : null,
-            accountId: account.id,
-            environmentId: environment.id,
-            tags: endUserPayload.tags || null
-        });
+    if (!connection.end_user_id) {
+        const createdEndUser = await createEndUser(db, { accountId: account.id, environmentId: environment.id, ...endUser });
         if (createdEndUser.isErr()) {
             return createdEndUser;
         }
 
         return Ok(createdEndUser.value);
     }
 
-    // --- Update
-    const endUser = endUserRes.value;
+    return await getAndUpdateEndUser(db, {
+        id: connection.end_user_id,
+        account,
+        environment,
+        endUser
+    });
+}
+
+async function getAndUpdateEndUser(
```

Comment:
would `updateIfModified` be a better name for this function?

Reply:
renamed to `getAndUpdateIfModifiedEndUser`